src/rust/core/server/src/threads/workers/storage.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use super::*;
+use crate::poll::Poll;
+use crate::QUEUE_RETRIES;
+use common::signal::Signal;
+use config::WorkerConfig;
+use core::time::Duration;
+use entrystore::EntryStore;
+use mio::{Events, Waker};
+use protocol::Execute;
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+/// A builder type for a storage worker which owns the storage and executes
+/// requests from a queue and returns responses back to the worker threads.
+pub struct StorageWorkerBuilder<Storage, Request, Response> {
+    poll: Poll,
+    nevent: usize,
+    timeout: Duration,
+    storage: Storage,
+    _request: PhantomData<Request>,
+    _response: PhantomData<Response>,
+}
+
+impl<Storage, Request, Response> StorageWorkerBuilder<Storage, Request, Response> {
+    /// Create a new `StorageWorkerBuilder` from the config and storage.
+    pub fn new<T: WorkerConfig>(config: &T, storage: Storage) -> Result<Self, std::io::Error> {
+        let poll = Poll::new().map_err(|e| {
+            error!("{}", e);
+            std::io::Error::new(std::io::ErrorKind::Other, "Failed to create epoll instance")
+        })?;
+
+        Ok(Self {
+            nevent: config.worker().nevent(),
+            timeout: Duration::from_millis(config.worker().timeout() as u64),
+            poll,
+            storage,
+            _request: PhantomData,
+            _response: PhantomData,
+        })
+    }
+
+    /// Returns the waker for the storage worker.
+    pub(crate) fn waker(&self) -> Arc<Waker> {
+        self.poll.waker()
+    }
+
+    /// Finalize the builder and return a `StorageWorker` by providing the
+    /// queues that are necessary for communication with other threads.
+    pub fn build(
+        self,
+        signal_queue: Queues<(), Signal>,
+        storage_queue: Queues<TokenWrapper<Option<Response>>, TokenWrapper<Request>>,
+    ) -> StorageWorker<Storage, Request, Response> {
+        StorageWorker {
+            poll: self.poll,
+            nevent: self.nevent,
+            timeout: self.timeout,
+            signal_queue,
+            storage: self.storage,
+            storage_queue,
+        }
+    }
+}
+
+/// A finalized `StorageWorker` which is ready to be run.
+pub struct StorageWorker<Storage, Request, Response> {
+    poll: Poll,
+    nevent: usize,
+    timeout: Duration,
+    signal_queue: Queues<(), Signal>,
+    storage: Storage,
+    storage_queue: Queues<TokenWrapper<Option<Response>>, TokenWrapper<Request>>,
+}
+
+impl<Storage, Request, Response> StorageWorker<Storage, Request, Response>
+where
+    Storage: Execute<Request, Response> + EntryStore,
+{
+    /// Run the `StorageWorker` in a loop, handling new session events.
+    pub fn run(&mut self) {
+        let mut events = Events::with_capacity(self.nevent);
+        let mut requests = Vec::with_capacity(1024);
```

Comment:
Do you intend to use `QUEUE_CAPACITY` instead of `1024`?

Reply:
Yes, thanks!